src/zenml/zen_server/cache.py

```diff
@@ -0,0 +1,195 @@
+#  Copyright (c) ZenML GmbH 2024. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Memory cache module for the ZenML server."""
+
+import time
+from collections import OrderedDict
+from threading import Lock
+from typing import Any, Callable, Dict, Optional
+from uuid import UUID
+
+from zenml.logger import get_logger
+from zenml.utils.singleton import SingletonMetaClass
+
+logger = get_logger(__name__)
+
+
+class MemoryCacheEntry:
+    """Simple class to hold cache entry data."""
+
+    def __init__(self, value: Any, expiry: int) -> None:
+        """Initialize a cache entry with value and expiry time.
+
+        Args:
+            value: The value to store in the cache.
+            expiry: The expiry time in seconds.
+        """
+        self.value: Any = value
+        self.expiry: int = expiry
+        self.timestamp: float = time.time()
+
+    @property
+    def expired(self) -> bool:
+        """Check if the cache entry has expired."""
+        return time.time() - self.timestamp >= self.expiry
+
+
+class MemoryCache(metaclass=SingletonMetaClass):
+    """Simple in-memory cache with expiry and capacity management.
+
+    This cache is thread-safe and can be used in both synchronous and
+    asynchronous contexts. It uses a simple LRU (Least Recently Used) eviction
+    strategy to manage the cache size.
+
+    Each cache entry has a key, value, timestamp, and expiry. The cache
+    automatically removes expired entries and evicts the oldest entry when
+    the cache reaches its maximum capacity.
+
+
+    Usage Example:
+
+        cache = MemoryCache()
+        uuid_key = UUID("12345678123456781234567812345678")
+
+        cached_or_real_object = cache.get_or_cache(
```

Comment:
I don't see the `get_or_cache` method anywhere

Reply:
Ah, you're right, this was left behind from an older implementation version.

Reply:
done